numpy/lib/_npyio_impl.py

```diff
@@ -1591,39 +1593,51 @@ def first_write(self, v):
         iscomplex_X = np.iscomplexobj(X)
         # `fmt` can be a string with multiple insertion points or a
         # list of formats.  E.g. '%10.5f\t%10d' or ('%10.5f', '$10d')
-        if type(fmt) in (list, tuple):
-            if len(fmt) != ncol:
-                raise AttributeError('fmt has wrong shape.  %s' % str(fmt))
-            format = delimiter.join(fmt)
-        elif isinstance(fmt, str):
-            n_fmt_chars = fmt.count('%')
-            error = ValueError('fmt has wrong number of %% formats:  %s' % fmt)
-            if n_fmt_chars == 1:
-                if iscomplex_X:
-                    fmt = [' (%s+%sj)' % (fmt, fmt), ] * ncol
-                else:
-                    fmt = [fmt, ] * ncol
+        if fstring_fmt:
+            if isinstance(fstring_fmt, str):
+                fstring_fmt = [fstring_fmt] * ncol
+            elif len(fstring_fmt) != ncol:
+                raise ValueError("fstring_fmt has wrong shape. %s" % str(fstring_fmt))
+            format_row = delimiter.join(fstring_fmt) + newline
+        else:
+            if isinstance(fmt, (list, tuple)):
+                if len(fmt) != ncol:
+                    raise AttributeError('fmt has wrong shape.  %s' % str(fmt))
                 format = delimiter.join(fmt)
-            elif iscomplex_X and n_fmt_chars != (2 * ncol):
-                raise error
-            elif ((not iscomplex_X) and n_fmt_chars != ncol):
-                raise error
+            elif isinstance(fmt, str):
+                n_fmt_chars = fmt.count('%')
+                error = ValueError('fmt has wrong number of %% formats:  %s' % fmt)
+                if n_fmt_chars == 1:
+                    if iscomplex_X:
+                        fmt = [' (%s+%sj)' % (fmt, fmt)] * ncol
+                    else:
+                        fmt = [fmt] * ncol
+                    format = delimiter.join(fmt)
+                elif iscomplex_X and n_fmt_chars != (2 * ncol):
+                    raise error
+                elif not iscomplex_X and n_fmt_chars != ncol:
+                    raise error
+                else:
+                    format = fmt
+                format += newline
             else:
-                format = fmt
-        else:
-            raise ValueError('invalid fmt: %r' % (fmt,))
+                raise ValueError('invalid fmt: %r' % (fmt,))
 
         if len(header) > 0:
             header = header.replace('\n', '\n' + comments)
-            fh.write(comments + header + newline)
-        if iscomplex_X:
+            fh.write(comments + header.rstrip('\n') + newline)
```

Comment:
This could change behaviour of the `fmt` argument (when `fstring_fmt` is not set). Do we need this change?

Reply:
No, we do not need this change. The modifications we discussed earlier are sufficient to address the issue of extra newline characters without altering the behavior of the fmt argument when fstring_fmt is not set. I tried to fix it for the newline errors I got during tests. I will change it back to original.

Reply:
This change is not necessary. I modified it for the newline errors I got when the tests were run. I will revert it to maintain the original behavior.